src/xrpld/app/tx/detail/SetTrust.cpp

```diff
@@ -182,6 +228,60 @@ SetTrust::preclaim(PreclaimContext const& ctx)
         }
     }
 
+    // Checking all freeze/deep freeze flag invariants.
+    if (ctx.view.rules().enabled(featureDeepFreeze))
+    {
+        bool const bNoFreeze = sle->isFlag(lsfNoFreeze);
+        bool const bSetFreeze = (uTxFlags & tfSetFreeze);
+        bool const bSetDeepFreeze = (uTxFlags & tfSetDeepFreeze);
+
+        if (bNoFreeze)
+        {
+            if (bSetFreeze || bSetDeepFreeze)
+            {
+                // Cannot freeze the trust line if NoFreeze is set
+                return tecNO_PERMISSION;
+            }
+        }
+        bool const bClearFreeze = (uTxFlags & tfClearFreeze);
+        bool const bClearDeepFreeze = (uTxFlags & tfClearDeepFreeze);
+        if ((bSetFreeze || bSetDeepFreeze) &&
+            (bClearFreeze || bClearDeepFreeze))
+        {
+            // Freezing and unfreezing in the same transaction should be
+            // illegal
+            return tecNO_PERMISSION;
+        }
+
+        bool const bHigh = id > uDstAccountID;
+        // Fetching current state of trust line
+        auto const sleRippleState =
+            ctx.view.read(keylet::line(id, uDstAccountID, currency));
+        std::uint32_t uFlags =
+            sleRippleState ? sleRippleState->getFieldU32(sfFlags) : 0u;
+        // Computing expected trust line state
+        uFlags = applyFreezeFlags(
+            uFlags,
+            bHigh,
+            bNoFreeze,
+            bSetFreeze,
+            bClearFreeze,
+            bSetDeepFreeze,
+            bClearDeepFreeze);
+
+        auto const frozen = uFlags & (bHigh ? lsfHighFreeze : lsfLowFreeze);
```

Comment:
I don't think this should be simplified.

In case one side frozen the trustline, the simplification of this check could allow another side to enact deep freeze (while not enacting normal freeze from his end). It is important to check flags of the same side of the trustline.

Reply:
The trustlines are bidirectional. Both sides can enact freeze technically on the same trustline.

Reply:
Wait, only the issuer can freeze the trustline, right?

Reply:
no, currency holder can also freeze the trust line, the effect is different though.

Reply:
Right, this is new in deep-freeze.

Reply:
not really, it's always been like this, but this is very obscure and kind of forgotten feature of the trust line